packages/unmock-node/src/backend/index.ts

```diff
@@ -24,58 +34,148 @@ async function handleRequestAndResponse(
+  private origOnSocket?: (socket: net.Socket) => void;
+  private mitm: any;
+
   public constructor(config?: INodeBackendOptions) {
     this.config = { ...nodeBackendDefaultOptions, ...config };
   }
 
   public initialize(options: UnmockOptions) {
-    mitm = Mitm();
-    mitm.on("connect", (socket: any, opts: any) => {
-      if (options.isWhitelisted(opts.host)) {
-        socket.bypass();
-      }
-    });
+    if (this.mitm !== undefined) {
+      this.reset();
+    }
+    this.mitm = Mitm();
+
+    const self = this;
+    this.origOnSocket = ClientRequest.prototype.onSocket;
+
+    /**
+     * When a socket is assigned to a client request, create an internal ID
+     * and add the ID in the request header. Thereby we can map
+     * the server-side incoming request to the corresponding
+     * client request and emit errors properly.
+     * Borrowed from https://github.com/FormidableLabs/yesno.
+     */
+    ClientRequest.prototype.onSocket = _.flowRight(
+      this.origOnSocket,
+      function(
+        this: ClientRequest,
+        socket: IBypassableSocket,
+      ): IBypassableSocket {
+        const requestId = uuidv4();
+        debugLog(
+          `New socket assigned to client request, assigned ID: ${requestId}`,
+        );
+        self.clientRequests[requestId] = { clientRequest: this };
+        this.setHeader(UNMOCK_INTERNAL_HTTP_HEADER, requestId);
+        return socket;
+      },
+    );
+
+    // Client-side socket connect, use to bypass connections
+    this.mitm.on("connect", (socket: IBypassableSocket, opts: RequestOptions) =>
+      this.mitmOnConnect.call(this, options, socket, opts),
+    );
+
     // Prepare the request-response mapping by bootstrapping all dependencies here
     const serviceDefLoader = new FsServiceDefLoader({
       servicesDir: this.config.servicesDirectory,
     });
-    const createResponse = responseCreatorFactory({ serviceDefLoader });
-    mitm.on("request", (req: IncomingMessage, res: ServerResponse) => {
-      handleRequestAndResponse(createResponse, req, res);
+    const createResponse = responseCreatorFactory({
+      serviceDefLoader,
     });
+
+    this.mitm.on("request", (req: IncomingMessage, res: ServerResponse) =>
+      this.mitmOnRequest.call(this, createResponse, req, res),
+    );
   }
+
   public reset() {
-    if (mitm) {
-      mitm.disable();
+    if (this.mitm) {
+      this.mitm.disable();
+      this.mitm = undefined;
+    }
+    if (this.origOnSocket) {
+      ClientRequest.prototype.onSocket = this.origOnSocket;
+      this.origOnSocket = undefined;
+    }
+    this.clientRequests = {};
+  }
+
+  private extractTrackedClientRequest(
+    incomingMessage: IncomingMessage,
+  ): ClientRequest {
+    const { [UNMOCK_INTERNAL_HTTP_HEADER]: reqId } = incomingMessage.headers;
+    debugLog(
+      `Intercepted incoming request with ID ${reqId}, matching to existing IDs:`,
+      Object.keys(this.clientRequests),
+    );
+    if (typeof reqId !== "string") {
```

Comment:
Is this the best way to check for non-empty request ID? I guess in theory `reqId` could be `""`?

Reply:
It's just for type-safety, headers are typed as  `string`, `undefined` or `string[]` and handling 
`string[]` is overkill here I think.